Subject: Feature: terminate-from
Origin: upstream, commit 1.10.0-7-g194620e
Author: Hrvoje Cavrak <hrvoje.cavrak@gmail.com>
Date: Tue Sep 8 17:38:24 2020 +1000

    To add a PPTP client to a Cisco VPDN group, terminate-from is often used
    to specify the hostname of the L2TP access concentrator or L2TP network
    server that will be required when accepting a VPDN tunnel.

    That hostname is part of a PPTP control packet header and in the case of
    Linux PPTP client, hardcoded to "local". This patch provides the ability
    to specify a different one through a command line argument.

--- a/pptp.c
+++ b/pptp.c
@@ -81,8 +81,8 @@
 int missing_window = MISSING_WINDOW;
 
 struct in_addr get_ip_address(char *name);
-int open_callmgr(struct in_addr inetaddr, char *phonenr, int argc,char **argv,char **envp, int pty_fd, int gre_fd);
-void launch_callmgr(struct in_addr inetaddr, char *phonenr, int argc,char **argv,char **envp);
+int open_callmgr(struct in_addr inetaddr, char *phonenr, char *client_hostname, int argc, char **argv,char **envp, int pty_fd, int gre_fd);
+void launch_callmgr(struct in_addr inetaddr, char *phonenr, char *client_hostname, int argc, char **argv,char **envp);
 int get_call_id(int sock, pid_t gre, pid_t pppd, 
 		 u_int16_t *call_id, u_int16_t *peer_call_id);
 void launch_pppd(char *ttydev, int argc, char **argv);
@@ -127,7 +127,8 @@
             "  --test-type <type>	Damage the packet stream by reordering\n"
             "  --test-rate <n>          Do the test every n packets\n"
             "  --missing-window <n>     Enable 'missing window' validation and set packet\n"
-            "                           tolerance (300=default, 6000=recommended)\n",
+            "                           tolerance (300=default, 6000=recommended)\n"
+            "  --client-hostname <name> Specify client hostname (used by Cisco VPDN Groups)\n",
 
             version, progname, progname);
     log("%s called with wrong arguments, program not started.", progname);
@@ -200,6 +201,8 @@
     char **pppdargv;
     char phonenrbuf[65]; /* maximum length of field plus one for the trailing
                           * '\0' */
+    char client_hostname_buf[64] = {0}; /* maximum length plus trailing '\0' */
+    char * volatile client_hostname = NULL;
     char * volatile phonenr = NULL;
     volatile int launchpppd = 1, debug = 0;
 
@@ -224,6 +227,7 @@
 	    {"rtmark", 1, 0, 0},
 	    {"nohostroute", 0, 0, 0},
 	    {"missing-window", 1, 0, 0},
+	    {"client-hostname", 1, 0, 0},
             {0, 0, 0, 0}
         };
         int option_index = 0;
@@ -330,7 +334,12 @@
 			    "and set to: %d\n", x);
 			missing_window = x;
 		    }
-                }
+		} else if (option_index == 18) { /* --client-hostname */
+		    strncpy(client_hostname_buf, optarg, sizeof(client_hostname_buf));
+		    client_hostname_buf[63] = '\0'; /* Terminate string to be safe */
+		    log("client-hostname set to %s\n", client_hostname_buf);
+		    client_hostname = client_hostname_buf;
+		}
                 break;
             case '?': /* unrecognised option */
                 /* fall through */
@@ -409,8 +418,8 @@
         /*
          * Open connection to call manager (Launch call manager if necessary.)
          */
-        callmgr_sock = open_callmgr(inetaddr, phonenr, argc, argv, envp,
-		pty_fd, gre_fd);
+        callmgr_sock = open_callmgr(inetaddr, phonenr, client_hostname, argc,
+		argv, envp, pty_fd, gre_fd);
         /* Exchange PIDs, get call ID */
     } while (get_call_id(callmgr_sock, parent_pid, child_pid, 
                 &call_id, &peer_call_id) < 0);
@@ -483,8 +492,8 @@
 }
 
 /*** start the call manager ***************************************************/
-int open_callmgr(struct in_addr inetaddr, char *phonenr, int argc, char **argv,
-        char **envp, int pty_fd, int gre_fd)
+int open_callmgr(struct in_addr inetaddr, char *phonenr, char *client_hostname,
+        int argc, char **argv, char **envp, int pty_fd, int gre_fd)
 {
     /* Try to open unix domain socket to call manager. */
     union {
@@ -517,7 +526,8 @@
                     /* close the pty and gre in the call manager */
                     close(pty_fd);
                     close(gre_fd);
-                    launch_callmgr(inetaddr, phonenr, argc, argv, envp);
+                    launch_callmgr(inetaddr, phonenr, client_hostname, argc,
+                        argv, envp);
                 }
                 default: /* parent */
                     waitpid(pid, &status, 0);
@@ -535,10 +545,10 @@
 }
 
 /*** call the call manager main ***********************************************/
-void launch_callmgr(struct in_addr inetaddr, char *phonenr, int argc __attribute__ ((unused)),
-        char**argv,char**envp) 
+void launch_callmgr(struct in_addr inetaddr, char *phonenr, char *client_hostname,
+        int argc __attribute__ ((unused)), char**argv, char**envp)
 {
-      char *my_argv[3] = { argv[0], inet_ntoa(inetaddr), phonenr };
+      char *my_argv[4] = { argv[0], inet_ntoa(inetaddr), phonenr, client_hostname };
       char buf[128];
       snprintf(buf, sizeof(buf), "pptp: call manager for %s", my_argv[1]);
 #ifdef PR_SET_NAME
--- a/pptp_callmgr.c
+++ b/pptp_callmgr.c
@@ -153,8 +153,9 @@
     signal(SIGPIPE, callmgr_do_nothing);
     signal(SIGUSR1, callmgr_do_nothing); /* signal state change
                                             wake up accept */
-    /* Step 2: Open control connection and register callback */
-    if ((conn = pptp_conn_open(inet_sock, 1, NULL/* callback */)) == NULL) {
+    /* Step 2: Open control connection and register callback
+     * argv[3] = client_hostname, will pass in header's hostname field */
+    if ((conn = pptp_conn_open(inet_sock, 1, argv[3], NULL/* callback */)) == NULL) {
         close(unix_sock); close(inet_sock); fatal("Could not open connection.");
     }
     FD_ZERO(&call_set);
--- a/pptp_ctrl.c
+++ b/pptp_ctrl.c
@@ -263,7 +263,8 @@
     
 
 /* Open new pptp_connection.  Returns NULL on failure. */
-PPTP_CONN * pptp_conn_open(int inet_sock, int isclient, pptp_conn_cb callback)
+PPTP_CONN * pptp_conn_open(int inet_sock, int isclient, char *client_hostname,
+        pptp_conn_cb callback)
 {
     PPTP_CONN *conn;
     /* Allocate structure */
@@ -299,6 +300,12 @@
             hton16(PPTP_MAX_CHANNELS), hton16(PPTP_FIRMWARE_VERSION), 
             PPTP_HOSTNAME, PPTP_VENDOR
         };
+
+        /* If a custom hostname value was passed, use that instead of
+         * default 'local'. */
+        if (client_hostname != NULL)
+            memcpy(packet.hostname, client_hostname, sizeof(packet.hostname));
+
         /* fix this packet, if necessary */
         int idx, rc;
         idx = get_quirk_index();
--- a/pptp_ctrl.h
+++ b/pptp_ctrl.h
@@ -24,7 +24,7 @@
  * not necessary if this is being opened by a server process after
  * receiving a conn_open packet from client. 
  */
-PPTP_CONN * pptp_conn_open(int inet_sock, int isclient, 
+PPTP_CONN * pptp_conn_open(int inet_sock, int isclient, char *client_hostname,
 			   pptp_conn_cb callback);
 PPTP_CALL * pptp_call_open(PPTP_CONN * conn, 
 			   pptp_call_cb callback, char *phonenr);
